moveit_ros/moveit_servo/src/utils/common.cpp

```diff
@@ -0,0 +1,333 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of the copyright holder nor the names of its
+ *   contributors may be used to endorse or promote products derived from
+ *   this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *******************************************************************************/
+
+/*      Title     : utils.cpp
+ *      Project   : moveit_servo
+ *      Created   : 05/17/2023
+ *      Author    : Andy Zelenak, V Mohammed Ibrahim
+ */
+
+#include <moveit_servo/utils/common.hpp>
+
+namespace
+{
+// The threshold above which `override_velocity_scaling_factor` will be used instead of computing the scaling from joint bounds.
+const double SCALING_OVERRIDE_THRESHOLD = 0.01;
+}  // namespace
+
+namespace moveit_servo
+{
+
+bool isValidCommand(const Eigen::VectorXd& command)
+{
+  // returns true only if there are no nan values.
+  return (command.array() == command.array()).all();
+}
+
+bool isValidCommand(const Eigen::Isometry3d& command)
+{
+  bool is_valid_rotation = true;
+  Eigen::Matrix3d identity, rotation;
+  identity.setIdentity();
+  rotation = command.linear();
+  // checks rotation, will fail if there is nan
+  is_valid_rotation = identity.isApprox(rotation.inverse() * rotation);
+  // Command is not vald if there is Nan
+  const bool not_nan = isValidCommand(command.translation());
+  return is_valid_rotation && not_nan;
+}
+
+bool isValidCommand(const TwistCommand& command)
+{
+  return !command.frame_id.empty() && isValidCommand(command.velocities);
+}
+
+bool isValidCommand(const PoseCommand& command)
+{
+  return !command.frame_id.empty() && isValidCommand(command.pose);
+}
+
+geometry_msgs::msg::Pose poseFromCartesianDelta(const Eigen::VectorXd& delta_x,
+                                                const Eigen::Isometry3d& base_to_tip_frame_transform)
+{
+  // get a transformation matrix with the desired position change &
+  // get a transformation matrix with desired orientation change
```

Comment:
Maybe make these comments less redundant, e.g.

```suggestion
  // Get separate transformation matrices with the desired position and orientation change.
```

Reply:
that was another suggestion I was considering. I'm fine with that too